pkg/sysched/sysched.go

```diff
@@ -0,0 +1,505 @@
+
+	for _, p := range pods {
+		syscall := sc.getSyscalls(p)
+		union(syscalls, syscall)
+	}
+
+	return syscalls
+}
+
+func (sc *SySched) removePod(pod *v1.Pod) {
+	nodeIP := pod.Status.HostIP
+
+	_, ok := sc.HostToPods[nodeIP]
+	if !ok {
+		klog.Infof("removePod: Host %s not yet cached", nodeIP)
+
+		return
+	}
+	for i, p := range sc.HostToPods[nodeIP] {
+		if p.Name == pod.Name {
+			sc.HostToPods[nodeIP] = remove(sc.HostToPods[nodeIP], i)
+			sc.HostSyscalls[nodeIP] = sc.recomputeHostSyscalls(sc.HostToPods[nodeIP])
+			c, _ := sc.getHostSyscalls(nodeIP)
+			klog.Info("remaining syscalls: ", c, " ", nodeIP)
+
+			return
+		}
+	}
+
+	return
+}
+
+/*
+func (sc *SySched) podAdded(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	klog.Infof("POD CREATED: %s/%s", pod.Namespace, pod.Name)
+}*/
+
+func (sc *SySched) podUpdated(old, new interface{}) {
+	pod := old.(*v1.Pod)
+	klog.Infof(
+		"POD UPDATED. %s/%s %s",
+		pod.Namespace, pod.Name, pod.Status.Phase,
+	)
+
+	if pod.Status.Phase == v1.PodSucceeded {
+		// finished
+		sc.removePod(pod)
+	} else if pod.Status.Phase == v1.PodPending && pod.Status.HostIP != "" {
+		//klog.Info("pod pending with IP assigned")
+		sc.addPod(pod)
+		//} else if (pod.Status.Phase == v1.PodRunning) {
+		//klog.Info("pod running")
+		//sc.addPod(pod)
+	} else if pod.Status.Phase == v1.PodFailed {
+		if pod.Status.Message != "" {
+			klog.Infof("pod failed: %s", pod.Status.Message)
+		}
+		klog.Infof("pod failed: pod.Status.Message unknown")
+	}
+}
+
+func (sc *SySched) podDeleted(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	klog.Infof("POD DELETED: %s/%s", pod.Namespace, pod.Name)
+	sc.removePod(pod)
+}
+
+// getArgs : returns the arguments for the SySchedArg plugin.
+func getArgs(obj runtime.Object) (*pluginconfig.SySchedArgs, error) {
+	SySchedArgs, ok := obj.(*pluginconfig.SySchedArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type SySchedArgs, got %T", obj)
+	}
+
+	return SySchedArgs, nil
+}
+
+// New initializes a new plugin and returns it.
+func New(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	sc := SySched{handle: handle}
+	sc.HostToPods = make(map[string][]*v1.Pod)
+	sc.HostSyscalls = make(map[string]map[string]bool)
+	//sc.CritSyscalls = make(map[string][]string)
+	sc.ExSAvg = 0
+	sc.ExSAvgCount = 1
+
+	args, err := getArgs(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the default syscall profile CR namespace and name for all syscalls
+	sc.DefaultProfileNamespace = args.DefaultProfileNamespace
+	sc.DefaultProfileName = args.DefaultProfileName
+
+	v1beta1.AddToScheme(scheme.Scheme)
+
+	clientSet, err := v1alpha1.NewForConfig(handle.KubeConfig())
+	if err != nil {
+		return nil, err
+	}
+
+	sc.clientSet = clientSet
+
+	podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+
+	//klog.Info("Setting up pod event handlers")
+	podInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			//AddFunc:    sc.podAdded,
```

Comment:
We only care when a pod has been assigned a node and is about to run on that node. This is when we update our hostsyscall map. This why we hook into the UpdateFunc to check if the pod has been assigned an IP to indicate that it is about to run.

Reply:
in that case, although hook it with UpdateFunc would also get what you want, but only AddFunc, you will be notified once, so much more efficient.

BTW: you can leverage the `FilterFunc` to filter out pods w/o `.spec.nodeName`:

https://github.com/kubernetes/kubernetes/blob/c682d2b8c59f615d6896a1b1db22a5e7a5aaa23e/pkg/scheduler/eventhandlers.go#L296-L312

Reply:
Maybe I understood when AddFunc gets called. I thought it gets called when a pod is introduced into the system and is about to be scheduled rather than when the pod gets added to a node to be run. 

Thanks for the FilterFunc pointer. That can be quite useful.